src/GafferCycles/IECoreCyclesPreview/Renderer.cpp

```diff
@@ -2680,108 +2688,42 @@ class CyclesRenderer final : public IECoreScenePreview::Renderer
+					if(
+						// e.g. "CPU" matches the first CPU device.
+						( typeIndex == 0 && StringAlgo::matchMultiple( typeString, devicePattern ) ) ||
+						// e.g. "CUDA:*" matches all CUDA devices, or `OPTIX:00` matches the first Optix device.
+						StringAlgo::matchMultiple( name, devicePattern )
+					)
 					{
-						if( deviceStr == g_defaultDeviceName.c_str() )
-						{
-							const auto device = m_deviceMap.find( g_defaultDeviceName );
-							if( device != m_deviceMap.end() )
-							{
-								m_multiDevices.push_back( device->second );
-
-								if( split.size() == 1 )
-								{
-									m_deviceName = g_defaultDeviceName;
-									break;
-								}
-								continue;
-							}
-						}
-						else if( boost::contains( deviceStr, ":" ) )
-						{
-							std::vector<std::string> _split;
-							boost::split( _split, deviceStr, boost::is_any_of( ":" ) );
-							ccl::DeviceType deviceType = ccl::Device::type_from_string( _split[0].c_str() );
-
-							if( _split[1] == "*" )
-							{
-								for( const auto &device : m_deviceMap )
-								{
-									if( deviceType == device.second.type )
-									{
-										m_multiDevices.push_back( device.second );
-									}
-								}
-							}
-							else
-							{
-								const auto device = m_deviceMap.find( deviceStr );
-								if( device != m_deviceMap.end() )
-								{
-									m_multiDevices.push_back( device->second );
-
-									if( split.size() == 1 )
-									{
-										m_deviceName = device->second.id;
-										break;
-									}
-									continue;
-								}
-								else
-								{
-									IECore::msg(
-										IECore::Msg::Warning, "CyclesRenderer::option",
-										fmt::format( "Cannot find device \"{}\" for option \"{}\".", deviceStr, name.string() )
-									);
-								}
-							}
-						}
-						else
-						{
-							IECore::msg(
-								IECore::Msg::Warning, "CyclesRenderer::option",
-								fmt::format( "Cannot find device \"{}\" for option \"{}\".", deviceStr, name.string() )
-							);
-						}
+						devices.push_back( device );
 					}
 				}
-				else
+
+				if( devices.empty() )
 				{
-					m_multiDevices.clear();
-					const auto device = m_deviceMap.find( g_defaultDeviceName );
-					if( device != m_deviceMap.end() )
-					{
-						m_multiDevices.push_back( device->second );
-					}
-					m_deviceName = g_defaultDeviceName;
-					IECore::msg( IECore::Msg::Warning, "CyclesRenderer::option", fmt::format( "Unknown value for option \"{}\".", name.string() ) );
+					IECore::msg( IECore::Msg::Warning, "CyclesRenderer", fmt::format( "No devices matching \"{}\" found, reverting to CPU.", devicePattern ) );
+					devices.push_back( firstCPUDevice() );
 				}
+
+				// Note : if there's only one device, `get_multi_device()` just
+				// returns it directly, rather than wrapping it.
+				m_sessionParams.device = ccl::Device::get_multi_device( devices, /* threads = */ 0, /* background = */ false );
```

Comment:
I suspect background needs to be `true` here. With this `false`, CPU devices aren't ever [returned](https://github.com/blender/cycles/blob/7d482abd0078765f52a49024c895a2f12a681cbd/src/device/device.cpp#L419-L422).

Then the second call to `get_multi_device` in `init()` would whittle this down once we know a final value for `background`?

Reply:
Good catch! Fixes in https://github.com/GafferHQ/gaffer/pull/5678/commits/b54a5c6d0a013798b5e6350a78413fd5f5faac2b.